### PR TITLE
Versions that don't specify all 4 parts are parsed incorrectly

### PR DESCRIPTION
**Pull request by tmat**
Adds missing test coverage of assembly and file version attributes.

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/roslyn-compiler @AlekseyTs Please review.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by cston** [comment]
LGTM

**Comment by tmat** [comment]
@AlekseyTs @cston Added another commit which adds additional test coverage of assemblies with wildcard assembly versions.

**Comment by tmat** [comment]
@dotnet-bot retest prtest/win/dbg/unit64 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/4576/
// Retest reason: https://github.com/dotnet/roslyn/issues/9335

**Comment by tmat** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/4601/
// Retest reason: https://github.com/dotnet/roslyn/issues/9181

File: src/Compilers/Core/Portable/VersionHelper.cs [hunk 1]
```diff
@@ -74,35 +74,50 @@ private static bool TryParse(string s, bool allowWildcard, ushort maxValue, out
...
-                        //alink would generate an error here saying "Cannot auto-generate build and 
-                        //revision version numbers for dates previous to January 1, 2000." Without
-                        //some refactoring here to relay the date problem, Roslyn
-                        //will generate an inaccurate error about the version string being of the wrong format.
-
-                        version = null;
-                        return false;
-                    }
-                    values[2] = (ushort)build;
+                    values[i] = ushort.MaxValue;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/VersionHelper.cs]: This behavior should probably be reflected in the doc comment for this method and probably in the doc comment of a public method that calls this helper.

File: src/Compilers/Core/Portable/VersionHelper.cs [hunk 2]
```diff
@@ -74,35 +74,50 @@ private static bool TryParse(string s, bool allowWildcard, ushort maxValue, out
 
             if (hasWildcard)
             {
-                int seconds = ((int)(DateTime.Now.TimeOfDay.TotalSeconds)) / 2;
-                if (seconds > (int)maxValue)
+                for (int i = lastExplicitValue; i < values.Length; i++)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/VersionHelper.cs]: Let's assert at the beginning of this function that, if allowWildcard is true, then maxValue <  ushort.MaxValue.

File: src/Compilers/Core/Portable/VersionHelper.cs [hunk 3]
```diff
@@ -74,35 +74,50 @@ private static bool TryParse(string s, bool allowWildcard, ushort maxValue, out
...
+            // (without taking into account time zone adjustments for daylight saving time), divided by 2."
+            if (time == default(DateTime))
+            {
+                time = DateTime.Now;
+            }
+
+            int revision = (int)time.TimeOfDay.TotalSeconds / 2;
+
+            // 24 * 60 * 60 / 2 = 43200 < 65534
+            Debug.Assert(revision < 0xffff);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/VersionHelper.cs]: ushort.MaxValue instead of 0xffff?

File: src/Compilers/Core/Portable/VersionHelper.cs [hunk 4]
```diff
@@ -74,35 +74,50 @@ private static bool TryParse(string s, bool allowWildcard, ushort maxValue, out
...
+
+            if (pattern.Build == ushort.MaxValue)
+            {
+                TimeSpan days = time.Date - new DateTime(2000, 1, 1);
+                int build = Math.Min(ushort.MaxValue, (int)days.TotalDays);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/VersionHelper.cs]: Didn't we say that ushort.MaxValue is an invalid value for a version part?

File: src/Compilers/Core/Portable/VersionHelper.cs (hunk 1 above)
Comment (tmat) [inline src/Compilers/Core/Portable/VersionHelper.cs]: Added.

File: src/Compilers/Core/Portable/VersionHelper.cs (hunk 2 above)
Comment (tmat) [inline src/Compilers/Core/Portable/VersionHelper.cs]: Added.

File: src/Compilers/Core/Portable/VersionHelper.cs (hunk 3 above)
Comment (tmat) [inline src/Compilers/Core/Portable/VersionHelper.cs]: Changed.

File: src/Compilers/Core/Portable/VersionHelper.cs (hunk 4 above)
Comment (tmat) [inline src/Compilers/Core/Portable/VersionHelper.cs]: It is not invalid. It's only considered invalid in the argument of AssemblyVersionAttribute. It's an artificial limitation imposed by the compiler (presumably native compilers did that to be able to represent *). Metadata allows version numbers [0, 0xffff].

**Comment by AlekseyTs** [comment]
LGTM

**Comment by tmat** [comment]
@dotnet-bot retest prtest/mac/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_mac_dbg_unit32/4579/
// Retest reason:

**Pull request merged**